plaid/src/benchmark/mod.rs

```diff
@@ -0,0 +1,168 @@
+    #[serde(default = "default_results_file_path")]
+    output_file_path: String,
+}
+
+/// Default file path for benchmarking results if none is provided in the config
+fn default_results_file_path() -> String {
+    format!(
+        "{}/benchmark_results/metrics.txt",
+        env!("CARGO_MANIFEST_DIR")
+    )
+}
+
+/// Metadata about a rule's execution
+pub struct ModulePerformanceMetadata {
+    /// The name of the module
+    pub module: String,
+    /// Time (in microseconds) for execution to complete
+    pub execution_time: u128,
+    /// The amount of computation used by the rule
+    pub computation_used: u64,
+}
+
+/// Represents a module's aggregate performance
+struct AggregatePerformanceData {
+    /// The number of times the module has been executed
+    runs: u64,
+    /// The total time (in microseconds) the module has spent in the execution loop
+    total_execution_time: u128,
+    /// The total computation used by the module
+    total_computation_used: u64,
+    /// Denotes whether the system should continue collecting performance metadata
+    maxed_out: bool,
+}
+
+impl AggregatePerformanceData {
+    /// Creates a new `AggregatePerformanceData` with initial execution time and computation used.
+    fn new(execution_time: u128, computation_used: u64) -> Self {
+        Self {
+            runs: 1,
+            total_execution_time: execution_time,
+            total_computation_used: computation_used,
+            maxed_out: false,
+        }
+    }
+
+    /// Updates the aggregate data atomically with the latest execution time and computation used.
+    /// If overflow occurs, the update is rolled back and none of the fields are modified.
+    fn update(&mut self, message: &ModulePerformanceMetadata) {
+        if self.maxed_out {
+            return;
+        }
+
+        // Check if the additions would overflow before proceeding
+        if let (Some(new_total_execution_time), Some(new_total_computation_used)) = (
+            self.total_execution_time
+                .checked_add(message.execution_time),
+            self.total_computation_used
+                .checked_add(message.computation_used),
+        ) {
+            // Atomic update
+            self.runs += 1;
+            self.total_execution_time = new_total_execution_time;
+            self.total_computation_used = new_total_computation_used;
+        } else {
+            error!("Overflow occurred updating execution data for [{}]. No further execution data will be collected.", message.module);
+            self.maxed_out = true;
+        }
+    }
+}
+
+impl Benchmarking {
+    pub async fn benchmark_loop(
+        &self,
+        receiver: Receiver<ModulePerformanceMetadata>,
+        shutdown_initiated: Arc<AtomicBool>,
+    ) {
+        let mut aggregate_performance_metadata = HashMap::new();
+
+        // Benchmarking loop runs until shutdown_initiated is set to true
+        while !shutdown_initiated.load(Ordering::SeqCst) {
```

Comment:
Is this needed? It seems to me that when a graceful shutdown occurs the sender for the below will be dropped causing looping to cease here anyway.

You also have a task here which has no awaits in it. This is basically going to pin an entire tokio thread. Should we not just await on the `recv` and let it shutdown by dropping the sender?

Reply:
Yeah I think it's probably possible to drop that channel and use it to trigger report generation.

And yeah, this pretty much pins the thread in its current state - will see if there's a way to make it less resource intensive